---
title: "iModMix -integrative Modules for Multi-omics data"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("knitr")

```

## Introduction

Welcome to the Integrative Modules for Multi-omics data (iModMix) Shiny app. This app implements integrative analysis of molecular data in a user-friendly interface to analyze:

- Metabolomics vs proteomics.
- Metabolomics vs transcriptomics.

<style>
  .github-button {
    position: absolute;
    top: 170px;
    right: 150px;
    border: 1px solid #ccc;
    padding: 10px;
    background-color: #f9f9f9;
    border-radius: 5px;
    text-align: center;
  }
</style>

<div class="github-button">
  <a href="https://github.com/IsisNarvaez/iModMix">
    <img src="https://github.com/favicon.ico" width="30" height="30" alt="GitHub" />
    <span>iModMix GitHub repository</span>
  </a>
</div>

## What is iModMix?
IModMix is a novel network-based approach that integrates multi-omics data, including metabolomics, proteomics, and transcriptomics data, into a unified network to unveil associations across various layers of omics data and reveal significant associations with phenotypes of interest. *(See Figure 1)

```{r, echo=F, fig.align="center", fig.width=2, out.width="100%", fig.cap="Figure 1: Overview of the iModMix's pipeline"}
knitr::include_graphics(paste0(here::here(),"/documentation/OverviewIModMix.png"), error = FALSE)
```


## What does this application do?

In general terms, this application will allow the user to:

- Generate a partial correlation network of all features available in the selected dataset.
- Interrogate this network to identify groups of features (modules) correlated with each other.
- Cross these modules with available clinical data.
- Explore the correlations between metabolite modules and protein/gene modules.
- Generate a network of modules between metabolomics and proteomics/transcriptomics datasets.



