---
title: "iModMix -integrative Modules for Multi-omics data"
date: "2024-04-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("knitr")

```

## Introduction

Welcome to the Integrative Modules for Multi-omics data (IModMix) shiny app. This app implements integrative analysis of molecular data in a user-friendly interface.

## What is IModMix?
IModMix is a novel integrative analysis framework by which networks and molecular modules are determined to increase the statistical power to uncover biological relevant associations with a phenotype of interest. 

## What does this application do?
This application allows users with little or no coding skills to try, explore and play with the IModMix method.

In general terms, this application will let the user:

generate a partial correlation network of all features available in the selected dataset,
interrogate this network to identify groups of genes (modules) correlated with each other,
cross these gene modules with available clinical data,
explore the relations between metabololites modules and gene modules and
generate a network of modules beetween metabolomics ans proteomics/genomics datasets. 

## Data input 

In the first section you can upload one or two omics tables
the metabolite,proteins and genes must be identified in, `Feture ID` which contains a unique ID for each row. *See Figure 1 for an example of the mass spec data input format.* 


```{r data-example, echo=F, fig.align="center", fig.width=2, out.width="50%", fig.cap="Figure 1: Example expression data that can be uploaded in the 'Metabolomics data' and 'Proteomics/Genomics data' tabs."}
knitr::include_graphics(paste0(here::here(),"/documentation/Expression.png"), error = FALSE)
```

Metabolomics data must be have `KEGG.ID` for the metabolites that have corresponding KEGG IDs for enrichment analysis.
Proteomics/Genomics data must be have `Symbol` for the features that have corresponding Symbols for enrichment analysis.
*See Figure 2 for an example of the annotation data input format.*

```{r, echo=F, fig.align="center", fig.width=2, out.width="50%", fig.cap="Figure 2: Example annotation data that can be uploaded in the 'Metabolomics data' and 'Proteomics/Genomics data' tabs."}
knitr::include_graphics(paste0(here::here(),"/documentation/Annotation.png"), error = FALSE)
```


The metadata files should have a column with the sample IDs that are used in the expression data files, identified in, `Sample`. This allows linking samples when having both positive and negative ion channel data in separate files. At least one additional column is required that groups samples together (think 'treatment' and 'control') that is selected for PCA, plots, etc.


```{r, echo=F, fig.align="center", fig.width=2, out.width="50%", fig.cap="Figure 3: Example metadata data that can be uploaded in the 'Metadata' tab."}
knitr::include_graphics(paste0(here::here(),"/documentation/Metadata.png"), error = FALSE)
```

